pkg/controllers/sharding/sharding_controller.go

```diff
@@ -0,0 +1,783 @@
+	vcinformers "volcano.sh/apis/pkg/client/informers/externalversions"
+	shardinformers "volcano.sh/apis/pkg/client/informers/externalversions/shard/v1alpha1"
+	shardlisters "volcano.sh/apis/pkg/client/listers/shard/v1alpha1"
+	"volcano.sh/volcano/pkg/controllers/framework"
+)
+
+const (
+	controllerName              = "sharding-controller"
+	defaultShardSyncPeriod      = 60 * time.Second
+	maxAssignmentCacheRetention = 5 * time.Minute
+	nodeCountChangeThreshold    = 0.5
+	updateTimeoutThreshold      = 15 * time.Second
+	nodeUsageChangeThreshold    = 0.5
+)
+
+func init() {
+	framework.RegisterController(&ShardingController{})
+}
+
+// ShardingController implements the framework.Controller interface
+type ShardingController struct {
+	ctx               context.Context
+	controllerOptions ShardingControllerOptions
+
+	kubeClient kubeclient.Interface
+	vcClient   vcclient.Interface
+
+	vcInformerFactory   vcinformers.SharedInformerFactory
+	kubeInformerFactory informers.SharedInformerFactory
+
+	nodeInformer  k8sinformerv1.NodeInformer
+	podInformer   k8sinformerv1.PodInformer
+	shardInformer shardinformers.NodeShardInformer
+
+	nodeLister  k8slisterv1.NodeLister
+	podLister   k8slisterv1.PodLister
+	shardLister shardlisters.NodeShardLister
+
+	nodeMetricsCache map[string]*NodeMetrics
+	metricsMutex     sync.RWMutex
+
+	queue          workqueue.TypedRateLimitingInterface[string]
+	nodeEventQueue workqueue.TypedRateLimitingInterface[string]
+
+	shardingManager  *ShardingManager
+	shardSyncPeriod  time.Duration
+	schedulerConfigs []SchedulerConfig
+
+	// assignment cache
+	assignmentCache *AssignmentCache
+	cacheMutex      sync.Mutex
+
+	// Event channels
+	assignmentChangeChan chan *AssignmentChangeEvent
+}
+
+// Return the name of the controller
+func (sc *ShardingController) Name() string {
+	return controllerName
+}
+
+// Initialize initializes the controller
+func (sc *ShardingController) Initialize(opt *framework.ControllerOption) error {
+	klog.V(6).Infof("Initializing ShardingController...")
+	sc.ctx = context.Background()
+	sc.controllerOptions = *NewShardingControllerOptions()
+
+	sc.kubeClient = opt.KubeClient
+	sc.vcClient = opt.VolcanoClient
+
+	sc.vcInformerFactory = opt.VCSharedInformerFactory
+	sc.kubeInformerFactory = opt.SharedInformerFactory
+
+	// Initialize informers
+	sc.nodeInformer = sc.kubeInformerFactory.Core().V1().Nodes()
+	sc.podInformer = sc.kubeInformerFactory.Core().V1().Pods()
+	sc.shardInformer = sc.vcInformerFactory.Shard().V1alpha1().NodeShards()
+
+	sc.nodeLister = sc.nodeInformer.Lister()
+	sc.podLister = sc.podInformer.Lister()
+	sc.shardLister = sc.shardInformer.Lister()
+
+	sc.initNodeIndices()
+
+	// Initialize queues
+	sc.queue = workqueue.NewTypedRateLimitingQueueWithConfig(
+		workqueue.DefaultTypedControllerRateLimiter[string](),
+		workqueue.TypedRateLimitingQueueConfig[string]{Name: controllerName},
+	)
+	sc.nodeEventQueue = workqueue.NewTypedRateLimitingQueueWithConfig(
```

Comment:
what does `node-events`? If this is for node, prefer naming `nodeQueue` and `nodeShardQueue` above